algorithm/cgal/ComputeToolpath.h

```diff
@@ -28,201 +34,265 @@ int ComputeToolpath(Geometry* geometry, size_t material, size_t selection,
-            continue;
-          }
-          if (const Point* point = boost::get<Point>(&intersection->first)) {
-            double p = CGAL::to_double(point->z().exact());
-            if (p > top) {
-              top = p;
-            }
-          }
-          if (const Segment* segment =
-                  boost::get<Segment>(&intersection->first)) {
-            double s = CGAL::to_double(segment->source().z().exact());
-            if (s > top) {
-              top = s;
-            }
-            double t = CGAL::to_double(segment->target().z().exact());
-            if (t > top) {
-              top = t;
-            }
+  // TODO: Apply a more sensible annealing strategy.
+  while (annealing > annealing_min) {
+    annealing_iteration += 1;
+    Cell* last_cell = nullptr;
+    Cell* current_cell = nullptr;
+    auto PickCutCell = [&](Cell* last_cell, Coord coord, double& next_cost,
+                           Cell*& next_cell) {
+      Cell* cell;
+      if (!GetCell(coord, cell)) {
+        return;
+      }
+      if (IsCellEligible(*cell)) {
+        double cost = ComputeCellCost(*cell);
+        if (last_cell) {
+          double distance = ComputeCoordDistance(last_cell->coord, coord);
+          if (distance < 1.9) {
+            // This cut is a 90 degree turn -- penalize it to reflect
+            // acceleration costs.
+            cost += kTurnPenalty;
           }
         }
-        int col = floor(x / tool_spacing);
-        int row = floor(y / tool_spacing);
-        std::pair<int, int> coord(col, row);
-        auto it = height_map.find(coord);
-        if (it == height_map.end() || it->second < top) {
-          height_map[coord] = top;
+        if (cost < next_cost) {
+          next_cost = cost;
+          next_cell = cell;
         }
       }
-    }
-  }
+    };
 
-  // This will give us back a height-map as a series of (possibly disconnected)
-  // square plates for visualization.
-  {
-    int target = geometry->add(GEOMETRY_MESH);
-    geometry->setIdentityTransform(target);
-    Surface_mesh& mesh = geometry->mesh(target);
-    Vertex_map vertices;
-
-    for (const auto& entry : height_map) {
-      const std::pair<int, int>& coord = entry.first;
-      const int x = coord.first;
-      const int y = coord.second;
-      const double height = entry.second;
-      mesh.add_face(
-          ensureVertex(
-              mesh, vertices,
-              Point((x - 0.0) * tool_spacing, (y - 0.0) * tool_spacing, height)),
-          ensureVertex(
-              mesh, vertices,
-              Point((x + 1.0) * tool_spacing, (y - 0.0) * tool_spacing, height)),
-          ensureVertex(
-              mesh, vertices,
-              Point((x + 1.0) * tool_spacing, (y + 1.0) * tool_spacing, height)),
-          ensureVertex(
-              mesh, vertices,
-              Point((x - 0.0) * tool_spacing, (y + 1.0) * tool_spacing, height)));
-    }
+    auto PickJumpCell = [&](Cell* current_cell, double& next_cost,
```

Comment:
This all looks good, but why is it in the .h file? I thought that .h files were only supposed to contain definitions? (Not complaining, just trying to learn)

Reply:
Just the usual header library approach.

One problem is that if we put this into a .cc file we may find that a lot of cgal gets duplicated between the various .o files coming out.

But it does make it slower to recompile.

Reply:
Ah right, I've asked this before. I remember now :)